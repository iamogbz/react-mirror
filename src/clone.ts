import { compare } from "specificity";
import {
    camelToSpinalCase,
    getPseudoElementSelector,
    splitSelectors,
} from "./utils";

export type CloneOptions = {
    styles?: React.CSSProperties;
    class?: string;
};

export function copyStyles(
    sourceElt: HTMLElement,
    targetElt: HTMLElement,
    options?: CloneOptions,
): void {
    const styleDecls: { [key: string]: CSSStyleDeclaration } = {};
    for (let i = 0; i < document.styleSheets.length; i++) {
        const styleSheet = document.styleSheets[i] as CSSStyleSheet;
        const rules = styleSheet.rules || styleSheet.cssRules;
        for (let j = 0; j < rules.length; j++) {
            const rule = rules[j] as CSSStyleRule;
            // test each selector in a group separately
            // accounts for bug with specificity library where
            // `.classA, .classB` fails when compared with `.classB`
            // https://github.com/keeganstreet/specificity/issues/4
            for (const selectorText of splitSelectors(rule.selectorText)) {
                // also match pseudo selectors
                const selector = selectorText.replace(
                    getPseudoElementSelector(selectorText),
                    "",
                );
                try {
                    if (sourceElt?.matches(selector || "*")) {
                        styleDecls[selectorText] = rule.style;
                    }
                } catch (e) {
                    // eslint-disable-next-line no-console
                    console.error(e.message, { selectorText });
                }
            }
        }
    }
    // ensures only the cloned styles are applied to element
    const singleClassName = targetElt.className.replace(" ", "-");
    targetElt.className = `${singleClassName}${options?.class || ""}`;
    // style element used for transfering pseudo styles
    const styleElt = document.createElement("style");
    styleElt.type = "text/css";
    for (const selector of Object.keys(styleDecls).sort(compare)) {
        const styleDecl = styleDecls[selector];
        // store pseudo element styles in style element
        const pseudoElementSelector = getPseudoElementSelector(selector);
        if (pseudoElementSelector) {
            styleElt.innerHTML += `.${targetElt.className}${pseudoElementSelector} { ${styleDecl.cssText} }\n`;
        } else {
            for (let i = 0; i < styleDecl.length; i++) {
                const propName = styleDecl[i];
                const propValue = styleDecl.getPropertyValue(propName);
                targetElt.style.setProperty(propName, propValue);
            }
        }
    }
    // append style element used for transfering pseudo styles
    styleElt.innerHTML && targetElt.appendChild(styleElt);
}

export function cloneNodeWithStyles(node: Node, options: CloneOptions): Node {
    const clone = node.cloneNode(false);
    // check that node is html element before cloning styles
    const cloneElt = clone as HTMLElement;
    const nodeElt = node as HTMLElement;
    if (nodeElt.style && cloneElt.style) {
        copyStyles(nodeElt, cloneElt, options);
        for (const [prop, value] of Object.entries(options.styles || {})) {
            cloneElt.style.setProperty(camelToSpinalCase(prop), value);
        }
    }
    return clone;
}

export function deepCloneWithStyles(node: Node, options: CloneOptions): Node {
    const clone = cloneNodeWithStyles(node, options);
    for (const child of node.childNodes) {
        clone.appendChild(deepCloneWithStyles(child, options));
    }
    return clone;
}
