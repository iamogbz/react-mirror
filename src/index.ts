export { Frame, FrameProps } from "./components/Frame";
export { Mirror, MirrorProps } from "./components/Mirror";
export { Reflection, ReflectionProps } from "./components/Reflection";
export { useMirror } from "./hooks/useMirror";
