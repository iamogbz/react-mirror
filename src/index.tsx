import * as React from "react";
import * as ReactDOM from "react-dom";

export function Mirror({
    reflect,
}: {
    reflect: React.ReactInstance | null;
}): JSX.Element {
    // ref to element in which reflection will be framed
    const [frame, ref] = React.useState<HTMLDivElement | null>(null);
    // real dom node of react element being reflected
    // eslint-disable-next-line react/no-find-dom-node
    const real = React.useMemo(() => ReactDOM.findDOMNode(reflect), [reflect]);
    // update the reflection to match the real node
    const update = React.useCallback(() => {
        if (!frame || !real) return;
        frame.innerHTML = "";
        frame.appendChild(real.cloneNode(true));
    }, [frame, real]);
    // start of the reflection
    React.useEffect(update, [update]);
    // mutation observer single instance
    const observer = React.useMemo(() => new MutationObserver(update), [
        update,
    ]);
    // observe or disconnect when real node is set or unmounted
    React.useEffect(() => {
        if (!real || !observer) return undefined;
        observer.observe(real, {
            attributes: true,
            childList: true,
            subtree: true,
            characterData: true,
        });
        return (): void => observer.disconnect();
    }, [real, observer]);
    // return frame element
    return <div ref={ref} style={{ pointerEvents: "none" }} />;
}

export function useMirror(): [
    React.Dispatch<unknown>,
    JSX.Element | undefined,
] {
    const [node, ref] = React.useState(null);
    return [ref, <Mirror key="mirror" reflect={node} />];
}
