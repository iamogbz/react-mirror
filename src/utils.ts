export function randomString(length: number): string {
    let str = "";
    while (str.length < length) {
        str += Math.random()
            .toString(36)
            .substr(2, length - str.length);
    }
    return str;
}

export function camelToSpinalCase(camelCase: string): string {
    return camelCase.replace(/([A-Z])/g, "-$1").toLocaleLowerCase();
}

export function splitSelectors(selectors: string): string[] {
    if (!selectors) {
        return [];
    }
    if (isAtRule(selectors)) {
        return [selectors];
    }
    const splitted = [];
    let quotes = 0;
    let parens = 0;
    let angulars = 0;
    let soFar = "";
    for (let i = 0, len = selectors.length; i < len; i++) {
        const char = selectors[i];
        if (char === "(") {
            parens += 1;
        } else if (char === ")") {
            parens -= 1;
        } else if (char === "[") {
            angulars += 1;
        } else if (char === "]") {
            angulars -= 1;
        } else if (char === '"') {
            quotes += 1;
        } else if (char === ",") {
            if (!parens && !angulars && !(quotes % 2)) {
                splitted.push(soFar.trim());
                soFar = "";
                continue;
            }
        }
        soFar += char;
    }
    splitted.push(soFar.trim());
    return splitted.filter(Boolean);
}

function isAtRule(selector: string): boolean {
    return selector.indexOf("@") === 0;
}

export function getPseudoElementSelector(selector: string): string {
    const pseudoElementSelector = /::\w+(\(\w+\))?/g;
    return selector.match(pseudoElementSelector)?.[0] ?? "";
}
