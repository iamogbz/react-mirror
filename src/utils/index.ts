export function randomString(length: number) {
    let str = "";
    while (str.length < length) {
        str += Math.random()
            .toString(36)
            .substr(2, length - str.length);
    }
    return str;
}

export function spinalToCamelCase(spinalCase: string): string {
    return spinalCase.replace(/-[a-z]/g, (m) => m.substring(1).toUpperCase());
}
