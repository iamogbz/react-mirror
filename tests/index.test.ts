import { helloWorld } from "../src";

describe("entry", (): void => {
    it("runs a test", (): void => {
        expect(helloWorld()).toMatchSnapshot();
    });
});
