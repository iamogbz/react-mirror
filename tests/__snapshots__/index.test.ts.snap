// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entry runs a test 1`] = `"hello world"`;
